proxy/envoy/agent.go

```diff
@@ -16,61 +16,71 @@ package envoy
 
 import (
 	"fmt"
-	"log"
 	"os"
 	"os/exec"
 	"sync"
+
+	"github.com/golang/glog"
 )
 
-// Agent manages a proxy service.
+// Agent manages a proxy instance
 type Agent interface {
-	Reload() error
+	Reload(config *Config) error
 }
 
 type agent struct {
-	config string            // Envoy config filename.
-	cmdMap map[*exec.Cmd]int // Map of known running Envoy processes and their restart epochs.
+	// Envoy binary path
+	binary string
+	// Map of known running Envoy processes and their restart epochs.
+	epoch  int
+	cmdMap map[*exec.Cmd]int
 	mutex  sync.Mutex
 }
 
+const (
+	EnvoyConfigPath   = "/etc/envoy/"
+	EnvoyFileTemplate = "envoy-rev%d.json"
+)
+
 // NewAgent creates a new instance.
-func NewAgent(config string) Agent {
+func NewAgent(binary string) Agent {
 	return &agent{
-		config: config,
+		binary: binary,
 		cmdMap: make(map[*exec.Cmd]int),
 	}
 }
 
+func configFile(epoch int) string {
+	return fmt.Sprintf(EnvoyConfigPath+EnvoyFileTemplate, epoch)
+}
+
 // Reload Envoy with a hot restart. Envoy hot restarts are performed by launching a new Envoy process with an
 // incremented restart epoch. To successfully launch a new Envoy process that will replace the running Envoy processes,
 // the restart epoch of the new process must be exactly 1 greater than the highest restart epoch of the currently
 // running Envoy processes. To ensure that we launch the new Envoy process with the correct restart epoch, we keep track
 // of all running Envoy processes and their restart epochs.
 //
 // Envoy hot restart documentation: https://lyft.github.io/envoy/docs/intro/arch_overview/hot_restart.html
-func (s *agent) Reload() error {
+func (s *agent) Reload(config *Config) error {
 	s.mutex.Lock()
 	defer s.mutex.Unlock()
 
-	// Find highest restart epoch of the known running Envoy processes.
-	restartEpoch := -1
-	for _, epoch := range s.cmdMap {
-		if epoch > restartEpoch {
-			restartEpoch = epoch
-		}
+	// Write config file
+	fname := configFile(s.epoch)
+	if err := config.WriteFile(fname); err != nil {
+		return err
 	}
-	restartEpoch++
 
 	// Spin up a new Envoy process.
-	cmd := exec.Command("envoy", "-c", s.config, "--restart-epoch", fmt.Sprint(restartEpoch))
+	cmd := exec.Command(s.binary, "-c", fname, "--restart-epoch", fmt.Sprint(s.epoch),
+		"--drain-time-s", "60", "--parent-shutdown-time-s", "90")
 	cmd.Stdout = os.Stdout
 	cmd.Stderr = os.Stderr
 	if err := cmd.Start(); err != nil {
 		return err
 	}
-
-	// Add the new Envoy process to the known set of running Envoy processes.
-	s.cmdMap[cmd] = restartEpoch
+	s.cmdMap[cmd] = s.epoch
```

Comment:
Thanks for pointing this out!
This was explained in the code but I was looking at the Envoy documentation which is not as explicit.
There is still a race condition possible if the latest Envoy dies in the middle of Reload().
I added a short delay to avoid stressing this race. I really think Envoy should relax that epoch+1 requirement; otherwise, I just don't see how to solve this race.